kernel/src/hil/flash.rs

```diff
@@ -12,14 +16,14 @@ pub trait Flash {
     /// when operations complete.
     fn set_client(&self, client: &'static Client);
 
-    /// Read data
-    fn read(&self, offset: usize, buf: &'static mut [u8]);
+    /// Read a page of flash.
+    fn read_page(&self, page_number: usize, buf: &'static mut [u8]) -> ReturnCode;
 
-    /// Write data
-    fn write(&self, offset: usize, buf: &'static mut [u8]);
+    /// Write a page of flash.
+    fn write_page(&self, page_number: usize, buf: &'static mut [u8]) -> ReturnCode;
```

Comment:
Yeah, I guess I didn't finish the example enough to illustrate why. An implementation might look like:

```rust
/// A non-existent flash hw controller with 16 byte pages
struct FlashController {
    // ... hw registers ...
}

impl Flash<[u8; 16]> for FlashController {
    ...
    fn write_page(&self, page_number: usize, &'static [u8; 16]) {
        ... // some implementation
    }
    ...
}
```

This would eliminate runtime checking because it would be impossible to pass anything except for a 16-byte array to `write_page`. In fact, technically the `AsMut` constraint isn't strictly necessary but it lets clients always convert whatever type is there into a byte-slice temporary without knowing the refified type (so I can write capsules that are generic across implementations and still do useful things).

An annoying downside I've elided is that implementations would probably have to create their own type since Rust only comes with `impl`s of `AsMut` for arrays up to like 24 bytes or something, and pages are much larger than that. It's not super hard:

```rust
pub struct MySpecificPage([u8; 4096]);

impl AsMut<[u8]> for MySpecificPage {
    fn as_mut(&self) -> &[u8] {
        &self.0
    }
}
```

And we could certainly provide implementations for common page sizes.